library/network_connections.py

```diff
@@ -1554,23 +1554,34 @@ def _complete_kwargs_loglines(self, rr, connections, idx):
             )
             prefix = prefix + (", '%s'" % (c["name"]))
         for r in rr["log"]:
-            yield (r[2], "[%03d] %s %s: %s" % (r[2], LogLevel.fmt(r[0]), prefix, r[1]))
+            yield (r[2], "[%03d] %s %s: %s" % (r[2], LogLevel.fmt(r[0]), prefix, r[1]), r[0])
 
-    def _complete_kwargs(self, connections, kwargs, traceback_msg=None):
+    def _complete_kwargs(self, connections, kwargs, traceback_msg=None, fail=False):
         if "warnings" in kwargs:
-            logs = list(kwargs["warnings"])
+            warning_logs = list(kwargs["warnings"])
         else:
-            logs = []
+            warning_logs = []
 
+        debug_logs = []
         loglines = []
         for res in self._run_results:
             for idx, rr in enumerate(res):
                 loglines.extend(self._complete_kwargs_loglines(rr, connections, idx))
         loglines.sort(key=lambda x: x[0])
-        logs.extend([x[1] for x in loglines])
+        for x in loglines:
+            if fail:
+                warning_logs.append(x[1])
+                if x[2] in (LogLevel.DEBUG, LogLevel.INFO):
+                    debug_logs.append(x[1])
+            else:
+                if x[2] in (LogLevel.DEBUG, LogLevel.INFO):
+                    debug_logs.append(x[1])
+                else:
+                    warning_logs.append(x[1])
         if traceback_msg is not None:
-            logs.append(traceback_msg)
-        kwargs["warnings"] = logs
+            warning_logs.append(traceback_msg)
```

Comment:
The initial issue was not so clear. I would prefer to:

- Return logs with warning or worse as warnings
- In case the module fails, return all logs as warnings
- Also return all logs (including warnings) in a new field. I was taking a look at [Ansibles commont return values](https://docs.ansible.com/ansible/latest/reference_appendices/common_return_values.html#common-return-values) and I guess we could use `stderr` for log messages. @pcahyna what do you think about this?

@elvgarrui what's your opinion?

Reply:
As far as I have seen this clarification makes sense, thanks!! Using standard names usually is the better option.

Reply:
I tried using `stderr` as new field but it seems like is not compatible. Instead, I could use `stderr_lines`, which is in the documentation said to keep the logs on a string format (as they are now). 

I'm not sure if this is the better option since the documentation says that the purpose of `stderr_lines` is, citing them: "When `stderr` is returned we also always provide this field which is a list of strings, one item per line from the original." - and there is no `stderr` field now, AFAIK.

The error that was returned when trying to assign the logs to stderr was: `AttributeError: 'list' object has no attribute 'splitlines'`

Reply:
`stderr` needs to be a string with the log lines separated by newlines, so `stderr = "\n".join(logs) + "\n"` should probably work.

Reply:
It works now, thanks!

Reply:
@tyll `stderr` is described as "Some modules execute command line utilities or are geared for executing commands directly (raw, shell, command, etc), this field contains the error output of these utilities.", I think you are abusing it, because this does not seem to be the case here.
Why do you need to return logs back actually? Is there a prior art for this in other Ansible modules? Who will read them in an automated setting? (Concerning warnings, see  https://blog.schlomo.schapiro.org/2015/04/warning-is-waste-of-my-time.html .) Can't you just send them to syslog on the managed host?
But if you really need to return them, I think you can just add another field to the result dict, instead of abusing stderr. Here is an example of a module that adds plenty of such fields: https://docs.ansible.com/ansible/latest/modules/authorized_key_module.html

Reply:
> @tyll `stderr` is described as "Some modules execute command line utilities or are geared for executing commands directly (raw, shell, command, etc), this field contains the error output of these utilities.", I think you are abusing it, because this does not seem to be the case here.

Do you see a specific problem with using this here? It does not conflict with any future development and stderr is meant to status output. To me it sounds like it should be intuitively clear for users what it means.

> Why do you need to return logs back actually? Is there a prior art for this in other Ansible modules? Who will read them in an automated setting? (Concerning warnings, see https://blog.schlomo.schapiro.org/2015/04/warning-is-waste-of-my-time.html .) Can't you just send them to syslog on the managed host?

These are valid questions. Personally, I found the info useful but that was not as a regular user. Nevertheless, this seems to be out of scope for the current issue at hand. I guess we can track/consider further enhancements. It seems that they could also be added to the common standard.

1) Consider removing the amount of log levels (this needs further discussion): I am not sure that I fully agree with the blog. IMHO warnings have also a place to tell about stuff that will become a problem/error in the future. Also, there might be things where we do not have a common agreement about how the tool should behave (like restarting NM)

2) Log to syslog, too. Sounds fine with me but can be added in a later step IMHO. Are there any recommendations/best practices to do this as part of Ansible?

> But if you really need to return them, I think you can just add another field to the result dict, instead of abusing stderr. Here is an example of a module that adds plenty of such fields: https://docs.ansible.com/ansible/latest/modules/authorized_key_module.html

What is the advantage of doing this?

Reply:
@pcahyna please take a look at my questions. I am fine with merging this once there are tests and am open to follow-up with further improvements if we can agree on them. Especialy syslog sounds fine just want to know more about it.